### ISSUE TITLE
Browserify Session

### ISSUE THREAD

**Issue by mikeal** [mikeal/nodeconf2014#16]
@substack thoughts on a browserify session/workshopper at NodeConf?

Given the breadth of the audience I'm wondering what could be done in terms of advanced challenges.

**Comment by ghost** [comment]
some advanced topics:
- custom transforms using the AST
- build your own browserify with module-deps and browser-pack
- [the compiler pipeline](https://github.com/substack/node-browserify/wiki/browserify-tools#wiki-compiler-pipeline)
- source maps (in transforms, [pulling apart](https://github.com/thlorenz/exorcist))
- using `--bare` and transforms to do crazy things with browsers or node (like [code coverage](https://github.com/substack/coverify))

**Comment by mikeal** [comment]
this is pretty exciting :)

**Comment by rvagg** [comment]
Node and browserify wizard @hughsk is really keen to help out on this so if you can find a way to pull in external help on this then he'd love that. (I'm working with him on another, project that has some overlaps here).

He's also interested in tackling the challenge of browserifying workshopper and the workshoppers themselves, which would be interesting.

**Comment by max-mapper** [comment]
we definitely need to browserify workshopper, perhaps via http://unix.substack.net/

also for beginners a workshopper about how to use browserify would be really cool! some ideas for challenges:
- your first bundle
- NPM scavenger hunt (produce expected output by finding modules on npm and combining them)
- sound synthesis (e.g. http://requirebin.com/#6110294)
- something visual, like http://requirebin.com/embed?gist=5941408
- drag + drop reading files https://github.com/maxogden/filereader-stream#example

basically show how to build complex things using a group of simple modules

**Comment by ghost** [comment]
I really like the scavenger hunt idea and more open-ended challenges. These would pair well with a leaderboard that you can browse once you've completed a level similarly to how project euler works.

**Comment by mikeal** [comment]
@rvagg since @substack is leading this session I'll let him pick who he'd like to help out and NodeConf will cover their expenses.

**Comment by mikeal** [comment]
how do you validate the sound synthesis? this is would be so damn fun.

**Comment by chrisdickinson** [comment]
I'd love to help with this.

Thoughts:
- Transform an existing require.js codebase to browserify.
- Use source transforms to generate rich error reporting (at work, we have a transform that takes the literal `"ua:trace"` and transforms that to report errors to Sentry with all of the variables in scope at that point)
- Transparently adding profiling, etc.
- Using glslify to connect the commonjs and webgl wizarding worlds, etc.

**Comment by chrisdickinson** [comment]
It'd also be neat to have a WebGL workshopper within the browserify workshopper (or other, web-games based workshopper), possibly as a "once you've finished the browserify workshopper, look at the neat things you can do with it!" sort of thing.

**Comment by mikeal** [comment]
I shutter to imagine what kind of code someone will have to write for the WebGL validator :)

**Comment by max-mapper** [comment]
IMO we should fly @mikolalysenko out and get him to write the 3D one, him and @chrisdickinson were talking about it in stackvm last night and it sounds really do-able, mikola has already been working on something similar https://github.com/mikolalysenko/gl-modules

**Comment by max-mapper** [comment]
Also for things that are hard to validate I think it is totally okay to describe what the output should be, or link to the working solution (but dont show the code) and have the verifications use the honor system

**Comment by mikeal** [comment]
hrm..... could this be its own session? if integrated with workshopper I don't think you'd have to have any prior knowledge about browserify. i think there is enough there that it could be its own session the only trick is that, because everything runs in parallel, it can't depend on any other session.

**Comment by chrisdickinson** [comment]
@mikeal yep, we could lean on something like beefy to make the browserify part of things transparent.

Oh man, I think I have an idea about how to verify webgl :|

**Comment by mikeal** [comment]
time to bring @mikolalysenko to see what he thinks :)

**Comment by mikolalysenko** [comment]
I actually started putting together a WebGL workshopper for gl-modules on Sunday.  I think we can get some sort of automatic verification working by then, though the biggest issue right now is to get @chrisdickinson's glslify integrated with the gl-modules stuff. I probably will take a look at creating the necessary browserify transforms this weekend or maybe sooner.

**Comment by mikeal** [comment]
ok, that settles it, I'll add this session to the schedule and start a new thread that is only for it.

**Comment by ghost** [comment]
https://github.com/substack/browserify-adventure
